File: Archive#01-101-200/152-Maximum Product Subarray/leetcode152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ret, tmp_p, tmp_n = nums[0], nums[0], nums[0]
        for elem in nums[1:]:
            positive = max(max(elem * tmp_p, elem * tmp_n), elem)
            negative = min(min(elem * tmp_p, elem * tmp_n), elem)
            ret = max(positive, ret)
            tmp_p, tmp_n = positive, negative
        return ret

File: Archive#01-101-200/152-Maximum Product Subarray/leetcode152.rb
# @param {Integer[]} nums
# @return {Integer}
def max_product(nums)
  ret, tmp_p, tmp_n = nums[0], nums[0], nums[0]
  nums[1..-1].each { |elem|
    positive = [[elem * tmp_p, elem * tmp_n].max, elem].max
    negative = [[elem * tmp_p, elem * tmp_n].min, elem].min
    ret = ret > positive ? ret : positive
    tmp_p, tmp_n = positive, negative
  }
  ret
end